src/shared/cgroup-setup.c

```diff
@@ -505,6 +506,67 @@ int cg_set_access(
         return 0;
 }
 
+int cg_set_access(
+                const char *controller,
+                const char *path,
+                uid_t uid,
+                gid_t gid) {
+        return cg_set_access_one(controller, path, uid, gid);
+}
+
+int cg_set_access_parents(
+                const char *controller,
+                const char *top,
+                const char *path,
+                uid_t uid,
+                gid_t gid) {
+        const char *p;
+        _cleanup_free_ char *prefix = NULL;
+        int r;
+
+        assert(top);
+
+        /* Set appropriate permissions for cgroups including parent cgroup.
+         * top is the root cgroup of a delegation subtree,
+         * path is a cgroup under the delegation subtree (can be equal to top).
+         */
+
+        r = path_extract_directory(top, &prefix);
+        if (r < 0)
+                return r;
+
+        /* make a copy for our mutations */
+        path = strdupa_safe(path);
+
+        p = path_startswith_full(path, prefix, false);
+        if (!p)
+                return -ENOTDIR;
+
+        /* path == top */
+        if (isempty(p))
+                p = path;
+
+        for (;;) {
+                char *s;
+                int n;
+
+                n = path_find_first_component(&p, false, (const char **)&s);
+                if (n <= 0)
+                        return n;
+
+                s[n] = '\0';
+                /* XXX paths below `top` should have access on all attributes */
+                r = cg_set_access_one(controller, path, uid, gid);
+                if (r < 0)
+                        return r;
+
+                s[n] = *p == '\0' ? '\0' : '/';
+        }
+
+        assert_not_reached();
+}
```

Comment:
i really don't like this. What do we gain by allowing full paths?

why isn't permitting a single subdir with no further depths enough?